File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/servers/list/ServerListAdapter.java
package com.lasthopesoftware.bluewater.client.servers.list;

import android.app.Activity;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.support.v4.content.LocalBroadcastManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.Button;
import android.widget.ImageButton;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.lasthopesoftware.bluewater.R;
import com.lasthopesoftware.bluewater.client.library.repository.Library;
import com.lasthopesoftware.bluewater.client.library.repository.LibrarySession;
import com.lasthopesoftware.bluewater.client.servers.list.listeners.EditServerClickListener;
import com.lasthopesoftware.bluewater.client.servers.list.listeners.SelectServerOnClickListener;
import com.lasthopesoftware.bluewater.client.servers.selection.BrowserLibrarySelection;
import com.lasthopesoftware.bluewater.client.servers.selection.IBrowserLibrarySelection;
import com.lasthopesoftware.bluewater.shared.android.view.ViewUtils;

import java.util.List;

public class ServerListAdapter extends BaseAdapter {

	private final List<Library> libraries;
	private final Library activeLibrary;
	private final IBrowserLibrarySelection browserLibrarySelection;
	private final Activity activity;

	private static class ViewHolder {
		final TextView textView;
		final Button btnSelectServer;
		final ImageButton btnConfigureServer;

		BroadcastReceiver broadcastReceiver;

		View.OnAttachStateChangeListener onAttachStateChangeListener;

		private ViewHolder(TextView textView, Button btnSelectServer, ImageButton btnConfigureServer) {
			this.textView = textView;
			this.btnSelectServer = btnSelectServer;
			this.btnConfigureServer = btnConfigureServer;
		}
	}

	public ServerListAdapter(Activity activity, List<Library> libraries, Library activeLibrary, IBrowserLibrarySelection browserLibrarySelection) {
		super();

		this.activity = activity;
		this.libraries = libraries;
		this.activeLibrary = activeLibrary;
		this.browserLibrarySelection = browserLibrarySelection;
	}

	@Override
	public View getView(final int position, View convertView, ViewGroup parent) {
		final Context parentContext = parent.getContext();
		final LocalBroadcastManager localBroadcastManager = LocalBroadcastManager.getInstance(parentContext);
		if (position == 0) {
			final RelativeLayout returnView = (RelativeLayout) getInflater(parentContext).inflate(R.layout.layout_standard_text, parent, false);
			final TextView textView = returnView.findViewById(R.id.tvStandard);
			textView.setText(parentContext.getText(R.string.btn_add_server));
			returnView.setOnClickListener(new EditServerClickListener(activity, -1));

			if (convertView != null && convertView.getTag() != null && ((ViewHolder)convertView.getTag()).broadcastReceiver != null)
				localBroadcastManager.unregisterReceiver(((ViewHolder) convertView.getTag()).broadcastReceiver);

			return returnView;
		}

		if (convertView == null || convertView.getTag() == null) {
			final RelativeLayout relativeLayout = (RelativeLayout) getInflater(parent.getContext()).inflate(R.layout.layout_server_item, parent, false);

			final TextView textView = relativeLayout.findViewById(R.id.tvServerItem);
			final Button btnSelectServer = relativeLayout.findViewById(R.id.btnSelectServer);
			final ImageButton btnConfigureServer = relativeLayout.findViewById(R.id.btnConfigureServer);

			relativeLayout.setTag(new ViewHolder(textView, btnSelectServer, btnConfigureServer));
			convertView = relativeLayout;
		}

		final ViewHolder viewHolder = (ViewHolder) convertView.getTag();
		final Library library = libraries.get(position - 1);
		viewHolder.textView.setText(library.getAccessCode());

		final Button btnSelectServer = viewHolder.btnSelectServer;

		viewHolder.textView.setTypeface(null, ViewUtils.getActiveListItemTextViewStyle(activeLibrary != null && library.getId() == activeLibrary.getId()));

		if (viewHolder.broadcastReceiver != null)
			localBroadcastManager.unregisterReceiver(viewHolder.broadcastReceiver);

		viewHolder.broadcastReceiver = new BroadcastReceiver() {
			@Override
			public void onReceive(Context context, Intent intent) {
				viewHolder.textView.setTypeface(null, ViewUtils.getActiveListItemTextViewStyle(library.getId() == intent.getIntExtra(LibrarySession.chosenLibraryInt, -1)));
			}
		};

		localBroadcastManager.registerReceiver(viewHolder.broadcastReceiver, new IntentFilter(BrowserLibrarySelection.libraryChosenEvent));

		if (viewHolder.onAttachStateChangeListener != null)
			parent.removeOnAttachStateChangeListener(viewHolder.onAttachStateChangeListener);

		viewHolder.onAttachStateChangeListener = new View.OnAttachStateChangeListener() {
			@Override
			public void onViewAttachedToWindow(View v) {

			}

			@Override
			public void onViewDetachedFromWindow(View v) {
				localBroadcastManager.unregisterReceiver(viewHolder.broadcastReceiver);
			}
		};

		parent.addOnAttachStateChangeListener(viewHolder.onAttachStateChangeListener);

		btnSelectServer.setOnClickListener(new SelectServerOnClickListener(library, browserLibrarySelection));

		viewHolder.btnConfigureServer.setOnClickListener(new EditServerClickListener(activity, library.getId()));

		return convertView;
	}

	private static LayoutInflater getInflater(Context context) {
		return (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
	}

	@Override
	public int getCount() {
		return libraries.size() + 1;
	}

	@Override
	public Object getItem(int position) {
		return libraries.get(position);
	}

	@Override
	public long getItemId(int position) {
		return position > 0 ? libraries.get(--position).getId() : -1;
	}

}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/servers/list/ServerListAdapter.kt
package com.lasthopesoftware.bluewater.client.servers.list

import android.app.Activity
import android.content.Context
import android.os.Handler
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageButton
import android.widget.RelativeLayout
import android.widget.TextView
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.lasthopesoftware.bluewater.R
import com.lasthopesoftware.bluewater.client.browsing.library.access.session.BrowserLibrarySelection
import com.lasthopesoftware.bluewater.client.browsing.library.access.session.SelectBrowserLibrary
import com.lasthopesoftware.bluewater.client.browsing.library.repository.Library
import com.lasthopesoftware.bluewater.client.servers.list.listeners.EditServerClickListener
import com.lasthopesoftware.bluewater.client.servers.list.listeners.SelectServerOnClickListener
import com.lasthopesoftware.bluewater.shared.android.adapters.DeferredListAdapter
import com.lasthopesoftware.bluewater.shared.android.view.LazyViewFinder
import com.lasthopesoftware.bluewater.shared.android.view.ViewUtils
import com.lasthopesoftware.bluewater.shared.android.view.getValue
import com.lasthopesoftware.bluewater.shared.messages.application.RegisterForApplicationMessages
import com.lasthopesoftware.bluewater.shared.messages.registerReceiver
import com.namehillsoftware.handoff.promises.Promise

class ServerListAdapter(private val activity: Activity, private val browserLibrarySelection: SelectBrowserLibrary, private val registerForApplicationMessages: RegisterForApplicationMessages)
	: DeferredListAdapter<Library, ServerListAdapter.ViewHolder>(activity, LibraryDiffer) {

	private val handler by lazy { Handler(activity.mainLooper) }
	private var activeLibrary: Library? = null

	fun updateLibraries(libraries: Collection<Library>, activeLibrary: Library?): Promise<Unit> {
		this.activeLibrary = activeLibrary
		return updateListEventually(libraries.toList())
	}

	override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
		val relativeLayout = getInflater(parent.context).inflate(R.layout.layout_server_item, parent, false) as RelativeLayout
		return ViewHolder(relativeLayout)
	}

	override fun onBindViewHolder(holder: ViewHolder, position: Int) {
		holder.update(getItem(position))
	}

	companion object {
		private fun getInflater(context: Context): LayoutInflater {
			return context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
		}
	}

	inner class ViewHolder(private val parent: RelativeLayout) : RecyclerView.ViewHolder(parent) {
		private val textView by LazyViewFinder<TextView>(parent, R.id.tvServerItem)
		private val btnSelectServer by LazyViewFinder<Button>(parent, R.id.btnSelectServer)
		private val btnConfigureServer by LazyViewFinder<ImageButton>(parent, R.id.btnConfigureServer)

		private var librarySelectionChangedSubscription: AutoCloseable? = null
		private var onAttachStateChangeListener: View.OnAttachStateChangeListener? = null

		fun update(library: Library) {
			textView.text = library.accessCode
			textView.setTypeface(null, ViewUtils.getActiveListItemTextViewStyle(library.id == activeLibrary?.id))

			librarySelectionChangedSubscription?.close()
			librarySelectionChangedSubscription = registerForApplicationMessages
				.registerReceiver(handler) { m: BrowserLibrarySelection.LibraryChosenMessage ->
					textView.setTypeface(
						null,
						ViewUtils.getActiveListItemTextViewStyle(library.id == m.chosenLibraryId.id)
					)
				}

			onAttachStateChangeListener?.also(parent::removeOnAttachStateChangeListener)
			parent.addOnAttachStateChangeListener(object : View.OnAttachStateChangeListener {
				override fun onViewAttachedToWindow(v: View) {}
				override fun onViewDetachedFromWindow(v: View) {
					librarySelectionChangedSubscription?.close()
				}
			}.apply { onAttachStateChangeListener = this })

			btnSelectServer.setOnClickListener(SelectServerOnClickListener(library, browserLibrarySelection))
			btnConfigureServer.setOnClickListener(EditServerClickListener(activity, library.id))
		}
	}

	private object LibraryDiffer : DiffUtil.ItemCallback<Library>() {
		override fun areItemsTheSame(oldItem: Library, newItem: Library): Boolean = oldItem.id == newItem.id

		override fun areContentsTheSame(oldItem: Library, newItem: Library): Boolean = areItemsTheSame(oldItem, newItem)
	}
}
